plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexExistingArtifactDeployCommandHandler.java

```diff
@@ -0,0 +1,83 @@
+
+import com.google.cloud.tools.eclipse.appengine.deploy.DeployPreferences;
+import com.google.cloud.tools.eclipse.appengine.deploy.StagingDelegate;
+import com.google.cloud.tools.eclipse.appengine.deploy.flex.FlexExistingArtifactDeployPreferences;
+import com.google.cloud.tools.eclipse.appengine.deploy.flex.FlexExistingDeployArtifactStagingDelegate;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployCommandHandler;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployPreferencesDialog;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.Messages;
+import com.google.cloud.tools.eclipse.googleapis.IGoogleApiFactory;
+import com.google.cloud.tools.eclipse.login.IGoogleLoginService;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import org.eclipse.core.commands.ExecutionEvent;
+import org.eclipse.core.commands.ExecutionException;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.swt.widgets.Shell;
+
+public class FlexExistingArtifactDeployCommandHandler extends DeployCommandHandler {
+
+  @Override
+  protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProject project,
+      IGoogleLoginService loginService, IGoogleApiFactory googleApiFactory) {
+    String title = Messages.getString("deploy.preferences.dialog.title.flexible");
+    return new FlexExistingArtifactDeployPreferencesDialog(shell, title, loginService,
+        googleApiFactory);
+  }
+
+  @Override
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
+    String appYamlPath = new FlexExistingArtifactDeployPreferences().getAppYamlPath();
+    IPath appYaml = resolveFileAgainstWorkspace(appYamlPath);
+    IPath appEngineDirectory = appYaml.removeLastSegments(1);
+
+    String deployArtifactPath = new FlexExistingArtifactDeployPreferences().getDeployArtifactPath();
+    IPath deployArtifact = resolveFileAgainstWorkspace(deployArtifactPath);
+
+    return new FlexExistingDeployArtifactStagingDelegate(deployArtifact, appEngineDirectory);
+  }
+
+  private IPath resolveFileAgainstWorkspace(String path) throws CoreException {
+    IPath fullPath = new Path(path);
+    if (!fullPath.isAbsolute()) {
+      IPath workspaceRoot = ResourcesPlugin.getWorkspace().getRoot().getLocation();
+      fullPath = workspaceRoot.append(path);
+    }
+
+    if (!fullPath.toFile().exists()) {
+      throw new CoreException(StatusUtil.error(this, fullPath + " does not exist."));
+    }
+    return fullPath;
+  }
+
+  @Override
+  protected DeployPreferences getDeployPreferences(IProject project) {
+    return new FlexExistingArtifactDeployPreferences();
+  }
+
+  @Override
+  protected IProject getSelectedProject(ExecutionEvent event)
+      throws ExecutionException, CoreException {
+    return null;
```

Comment:
Should this be throw new UnsupportedOperationException?
Either way, this does make me wonder if the abstractions are correct. I.e. should DeployCommandHandler have a getProject method when not all deploy commands have projects?

Reply:
Yeah, I also see the abstractions are somewhat weird with the introduction of the global dialog, which doesn't required a project. That's a gripe here, but unfortunately, the existing codebase assumed there is always a project to be deployed, and getting away from it with the global dialog requires massive refactoring. I tried some ideas, but it has been tricky when also considering #2416. Let me file a bug for having better abstractions.

BTW, we need to return `null` for now.

Reply:
Tech debt: #2449